src/libraries/drand/_index.md

```diff
@@ -0,0 +1,119 @@
+
+- that the `Signature` field is verified by the beacon's `PublicKey` as the beacon's signature of `SHA256(PreviousSignature || Round)`.
+- that the `Randomness` field is `SHA256(Signature)`.
+
+See [drand](https://github.com/drand/drand/blob/master/beacon/beacon.go#L63) for an example.
+
+#### Fetching the appropriate drand value while mining
+
+There is a deterministic mapping between a needed drand round number and a Filecoin epoch number.
+
+After initializing access to a drand beacon, a Filecoin node should have access to the following values:
+
+- `filEpochDuration`    -- the Filecoin network's epoch duration (between any two leader elections)
+- `filGenesisTime`      -- the Filecoin genesis timestamp
+- `filEpoch`            -- the current Filecoin epoch
+- `drandGenesisTime`    -- drand's genesis timestamp
+- `drandPeriod`         -- drand's epoch duration (between any two randomness creations)
+
+Using the above, a Filecoin node can determine the appropriate drand round value to be used for use in {{<sref leader_election>}} in an epoch using both networks' reliance on real time as follows:
+
+```go
+MaxBeaconRoundForEpoch(filEpoch) {
+    // determine what the latest Filecoin timestamp was from the current epoch number
+    var latestTs
+    if filEpoch == 0 {
+        latestTs = filGenesisTime
+    } else {
+        latestTs = ((uint64(filEpoch) * filEpochDuration) + filGenesisTime) - filEpochDuration
+    }
+    // determine the drand round number corresponding to this timestamp
+    // keeping in mind that drand emits round 1 at the drandGenesisTime
+    dround := (latestTs - drandGenesisTime) / uint64(drandPeriod) + 1
+    return dround
+}
+```
+
+### Edge cases and dealing with a drand outage
+
+It is important to note that any drand beacon outage will effectively halt Filecoin block production. Given that new randomness is not produced, Filecoin miners cannot generate new blocks.
+
+After a beacon downtime, drand nodes will work to quickly catch up to the current round, as defined by wall clock time. In this way, the above time-to-round mapping in drand (see above) used by Filecoin remains an invariant after this catch-up following downtime.
+
+So while Filecoin miners were not able to mine during the drand outage, they will quickly be able to run leader election thereafter, given a rapid production of drand values. We call this a "catch up" period.
+
+During the catch up period, Filecoin nodes will backdate their blocks in order to continue using the same time-to-round mapping to determine which drand round should be integrated according to the time. Miners can then choose to publish their null blocks for the outage period (including the appropriate drand entries throughout the blocks, per the time-to-round mapping), or (as is more likely) try to craft valid blocks that might have been created during the outage.
```

Comment:
Some comments:
1.
> Miners can then choose to publish their null blocks for the outage period 

Not clear - we can't publish a null block because a null block doesn't exist. 
2. Do we agree that one can "simply" append all the "catched-up" drand entries in the next block he's creating, right ? I believe the function below `ValidateBeaconEntries(blockHeader, priorBlockHeader) error {` does allow for that.

Reply:
>Miners can then choose to publish their null blocks for the outage period (including the appropriate drand entries throughout the blocks, per the time-to-round mapping)

Even though I don't quite understand this, this is not a behavior that Lotus codes at the moment (which is doing the "catch up" as well, mine as fast as possible).

Reply:
>  2. Do we agree that one can "simply" append all the "catched-up" drand entries in the next block he's creating, right ? I believe the function below ValidateBeaconEntries(blockHeader, priorBlockHeader) error { does allow for that.

I agree with that, on 1, what I mean is publish blocks, which will implicitly publish the null blocks. This is the key question, after an outage should we have a deterministic block/entry schedule (ie we know exactly how many entries should fit into the series of blocks and how they will be mapped onto those blocks), or not (ie there could be many valid drand/block mappings published on recovery).

I thought it was the latter @nicola , though you are right @nikkolasg that lotus implements one of those versions (ie catch up as fast as you can). IMO, that is not in conflict with the spec as written.